ddtrace/tracer.py

```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
         :param dict tags: dict of tags to set at tracer level
         """
         self.tags.update(tags)
+
+    def on(self, hook_name):
+        """
+        Decorator used to configure a tracer span hook
+
+        Usage::
+
+            # Register a hook for any emitted `my.span` Spans
+            @tracer.on('my.span')
+            def on_falcon_request(span, data):
+                span.set_tag('key', data['key'])
+
+
+            # Create and emit a `my.span` Span with the provided data
+            with tracer.trace('my.span') as span:
+                tracer.emit(span, dict(key='value'))
+
+        :param hook_name: The name of the hook to configure a hook for
+        :type hook_name: str
+        :rtype: function
+        :returns: A function decorator that registers a function for the provided hook name
+        """
+        def wrapper(func):
+            self.hooks[hook_name].add(func)
+            return func
+        return wrapper
+
+    def emit(self, span, *args, **kwargs):
+        """
+        Call all registered span hooks for the provided :class:`ddtrace.span.Span`
+
+        Usage::
+
+            # Register a hook for any emitted `my.span` Spans
+            @tracer.on('my.span')
+            def on_falcon_request(span, data):
+                span.set_tag('key', data['key'])
+
+
+            # Create and emit a `my.span` Span with the provided data
+            with tracer.trace('my.span') as span:
+                tracer.emit(span, dict(key='value'))
+
+        :param span: The span to emit a hook for
+        :type span: :class:`ddtrace.span.Span`
+        :param *args: positional arguments to pass to hook functions
+        :param **kwargs: keyword arguments to pass to hook functions
+        :rtype: None
+        :raises: TypeError
+        """
+        if not isinstance(span, Span):
+            raise TypeError('Tracer.emit expected a `ddtrace.span.Span` instead got "{}"'.format(span))
```

Comment:
In general we should never fail so exceptions must never bubble up to users' code. Because this is used by our instrumentation (and we cover it with unittests / integration tests), it's fine.